controllers/openstackcluster_controller.go

```diff
@@ -226,6 +229,13 @@ func deleteBastion(scope scope.Scope, cluster *clusterv1.Cluster, openStackClust
 		return err
 	}
 
+	// We must adopt the dependent resources before attempting to delete them. This is because
+	// adoption relies on other resources in the status, like networking.
+	err = compute.ResolveDependentBastionResources(scope, openStackCluster, bastionName(cluster.Name))
```

Comment:
There seems to be a bug here, The ResolveDependentBastionResources runs when no bastion is specified (In my case =nil).

This occurs during cluster creation meaning no bastion have ever existed. I assume there need to be some kind of check that verifies if there have ever been a bastion prior to running delete (on row 360 in this file, https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1788/files#diff-39b99bc838c9f76967790a10572cc44f0c41ccacddcd6e1eff934828d4b050a3R359-R361)

Stacktrace:
```log
I0213 14:58:13.426777       1 controller.go:115] "Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference" controller="openstackcluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="OpenStackCluster" OpenStackCluster="default/hux-lab1" namespace="default" name="hux-lab1" reconcileID="3bd08046-af33-4d4c-90a6-8cd063dba2d6"
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x80 pc=0x1afd40c]

goroutine 417 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:116 +0x1e5
panic({0x1d24520?, 0x3548260?})
	/usr/local/go/src/runtime/panic.go:914 +0x21f
sigs.k8s.io/cluster-api-provider-openstack/pkg/cloud/services/compute.ResolveDependentBastionResources({0x23934d8, 0xc000e241c0}, 0xc000f00c00, {0xc001284ef0, 0x10})
	/workspace/pkg/cloud/services/compute/dependent_resources.go:57 +0x12c
sigs.k8s.io/cluster-api-provider-openstack/controllers.deleteBastion({0x23934d8, 0xc000e241c0}, 0xc000e9c1a0, 0xc000f00c00)
	/workspace/controllers/openstackcluster_controller.go:234 +0x14a
sigs.k8s.io/cluster-api-provider-openstack/controllers.reconcileBastion({0x23934d8, 0xc000e241c0}, 0xc000e9c1a0, 0xc000f00c00)
	/workspace/controllers/openstackcluster_controller.go:360 +0x5d8
sigs.k8s.io/cluster-api-provider-openstack/controllers.reconcileNormal({0x23934d8, 0xc000e241c0}, 0x0?, 0xc000f00c00)
	/workspace/controllers/openstackcluster_controller.go:324 +0x11e
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackClusterReconciler).Reconcile(0xc0005340c0, {0x238a498?, 0xc000c0e1e0}, {{{0xc000d5a508?, 0x0?}, {0xc000d5a500?, 0xc001172d48?}}})
	/workspace/controllers/openstackcluster_controller.go:139 +0x714
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x238e848?, {0x238a498?, 0xc000c0e1e0?}, {{{0xc000d5a508?, 0xb?}, {0xc000d5a500?, 0x0?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:119 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00051e280, {0x238a4d0, 0xc000353950}, {0x1de9760?, 0xc0004041a0?})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:316 +0x3cc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00051e280, {0x238a4d0, 0xc000353950})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:266 +0x1c9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:227 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 234
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:223 +0x565
```

Reply:
nice spot, I'll have a look.

Reply:
on which file? the link didn't work for me. We have the following check:

```golang
	// Resolve and store referenced & dependent resources for the cluster
	if openStackCluster.Spec.Bastion != nil && openStackCluster.Spec.Bastion.Enabled {
		if openStackCluster.Status.Bastion == nil {
			openStackCluster.Status.Bastion = &infrav1.BastionStatus{}
		}
		changed, err := compute.ResolveReferencedMachineResources(scope, openStackCluster, &openStackCluster.Spec.Bastion.Instance, &openStackCluster.Status.Bastion.ReferencedResources)
		if err != nil {
			return reconcile.Result{}, err
		}
		if changed {
			// If the referenced resources have changed, we need to update the OpenStackCluster status now.
			return reconcile.Result{}, nil
		}

		changed, err = compute.ResolveDependentBastionResources(scope, openStackCluster, bastionName(cluster.Name))
		if err != nil {
			return reconcile.Result{}, err
		}
		if changed {
			// If the dependent resources have changed, we need to update the OpenStackCluster status now.
			return reconcile.Result{}, nil
		}
	}
```

So I'm surprised you enter in the if, since you have no bastion defined.

Reply:
I'm convinced we need a Bastion E2E test.